### DISCUSSION TITLE
Variant schema works but TypeScript says there's something wrong

### DISCUSSION THREAD

**Discussion by brandonpittman** [open-circle/valibot#269] [Q&A] [answered]
Can anyone tell me what's the problem here?

https://stackblitz.com/edit/typescript-pc279k?file=index.ts

If you look at this example, I'm validating prefecture and cities in Japan. My variant code works, but TypeScript says there's a problem and I think because of this error, I can't use `Input<typeof XXX>` to infer a type.

**Comment by fabian-hiller** [comment]
I think the problem here is that `variant` expects a tuple with at least two items. Since `.map()` types `locationsArray` as an array and not as a tuple, this probably means that the typing of `variant` does not accept it as an argument. As a workaround, you can use type casting. However, we should find a better solution for this in the future.

```ts
const locations = [
  { pref: '北海道', city: '札幌市' },
  { pref: '北海道', city: '函館市' },
  // ...
] as const;

type LocationItem = ObjectSchema<{
  pref: LiteralSchema<(typeof locations)[number]['pref']>;
  city: LiteralSchema<(typeof locations)[number]['city']>;
}>;

const locationsArray = locations.map((l) =>
  object({ pref: literal(l.pref), city: literal(l.city) })
) as [LocationItem, LocationItem, ...LocationItem[]];

export const LocationSchema = variant('pref', locationsArray);
```

Or

```ts
const locations = [
  { pref: '北海道', city: '札幌市' },
  { pref: '北海道', city: '函館市' },
  // ...
] as const;

const locationsArray = locations.map((l) =>
  object({ pref: literal(l.pref), city: literal(l.city) })
);

type LocationItem = (typeof locationsArray)[number];

export const LocationSchema = variant(
  'pref',
  locationsArray as [LocationItem, LocationItem, ...LocationItem[]]
);
```

**Reply by brandonpittman** [reply]
Types are fine, but I'm seeing an issue what I ran into after my original post:

Validation passes for the first prefecture/city pair, but not for any of the other ones.

**Reply by brandonpittman** [reply, marked as answer]
This doesn't work how I thought it would. Since the type is "pref", `variant` assumes there's only one `pref`, right? So only the first `pref`/`city` pair is considered valid.

This should actually be some sort of union:

`[{pref: "a", city: 1}, {pref: "a", city: 2}, [{pref: "b", city: 3} /* ... */]`

**Reply by brandonpittman** [reply]
Fixed it, changing `variant` to `union` worked. Sorry.
